scalafix/rules/src/main/scala/fix/Scalaz2cats_v1_0_0.scala

```diff
@@ -223,3 +223,71 @@ case class RemoveGlobalImports(index: SemanticdbIndex)
     }.asPatch
   }
 }
+
+case class MigrateEqual(index: SemanticdbIndex) extends SemanticRule(index, "MigrateEqual") {
+
+  private lazy val scalazEqualSyntaxImport = importer"scalaz.syntax.equal._"
+  private lazy val scalazEqualImport = importer"scalaz.Equal"
+  private lazy val catsEqualSyntaxImport = importer"cats.syntax.eq._"
+  private lazy val catsEqualImport = importer"cats.Eq"
+
+  private lazy val Equal = SymbolMatcher.normalized(
+    Symbol("_root_.scalaz.Equal.")
+  )
+
+  private lazy val catsEqual = SymbolMatcher.normalized(
+    Symbol("_root_.cats.Eq.")
+  )
+
+  private lazy val `===` = SymbolMatcher.normalized(
+    Symbol("_root_.scalaz.syntax.EqualOps.`===`.")
+  )
+  private lazy val `=/=` = SymbolMatcher.normalized(
+    Symbol("_root_.scalaz.syntax.EqualOps.`=/=`.")
+  )
+
+  private lazy val `/==` = SymbolMatcher.normalized(
+    Symbol("_root_.scalaz.syntax.EqualOps.`/==`.")
+  )
+
+  private lazy val `≠` = SymbolMatcher.normalized(
+    Symbol("_root_.scalaz.syntax.EqualOps.`≠`.")
+  )
+
+  private lazy val `≟` = SymbolMatcher.normalized(
+    Symbol("_root_.scalaz.syntax.EqualOps.`≟`.")
+  )
+
+  override def fix(ctx: RuleCtx): Patch = {
+    ctx.tree.collect {
+      case t @ importer"scalaz.syntax.equal._" =>
+        ctx.replaceTree(t, catsEqualSyntaxImport.syntax)
+      case t @ importer"scalaz.Equal" =>
+        ctx.replaceTree(t, catsEqualImport.syntax)
+      case t @ Type.Apply(Equal(_), List(tpe)) =>
+        ctx.replaceTree(t, q"Eq[$tpe]".syntax)
+      case Term.Select(a @ Equal(_), _) =>
+        ctx.replaceTree(a, q"Eq".syntax)
+      case Type.Param(_, _, _, _, _, List(a @ Equal(_))) =>
+        ctx.replaceTree(a, q"Eq".syntax)
+    }.asPatch + ctx.replaceSymbols(
+      "scalaz.syntax.EqualOps.`=/=`" -> "`=!=`",
+      "scalaz.syntax.EqualOps.`/==`" -> "`=!=`",
+      "scalaz.syntax.EqualOps.`≠`" -> "`=!=`",
+      "scalaz.syntax.EqualOps.`≟`" -> "`===`",
+      "scalaz.Equal.equalA" -> "fromUniversalEquals",
+      "scalaz.Equal.equalBy" -> "by",
+      "scalaz.Equal.equal" -> "eqv"
+    ) + (if (ctx.tree.collect {
+             case Term.ApplyInfix(_, `===`(_) | `=/=`(_) | `/==`(_) | `≠`(_) | `≟`(_), _, _)
+                 if !ctx.tree.contains(scalazEqualSyntaxImport) =>
+               ()
+           }.nonEmpty) ctx.addGlobalImport(catsEqualSyntaxImport)
+         else Patch.empty) + (if (ctx.tree.collect {
```

Comment:
I would move this last two patches into a separate private method, along the lines of `def addSyntaxImportIfNeeded(ctx: RuleCtx): Patch` or something like that.

The strategy I've taken elsewhere is to use `collectFirst`, `map` over it, and invoke `.asPatch` (which returns `Patch.empty` on a `None`). If you grep for `collectFirst` you'll find at least one example of what I mean.

Reply:
@gabro I'm going to add below method to `ScalafixUtils` class. It can be used for further migration rules
```
  def addImportsIfNeeded(
        ctx: RuleCtx
    )(checker: PartialFunction[Tree, Unit], importCheck: Importer, importAdd: Importer): Patch = {
      ctx.tree
        .collectFirst(checker)
        .filter(_ => !ctx.tree.contains(importCheck))
        .map(_ => ctx.addGlobalImport(importAdd))
        .asPatch
    }
```

Reply:
Nice idea 👍